crates/valence_network/src/connect.rs

```diff
@@ -375,36 +377,52 @@ fn auth_digest(bytes: &[u8]) -> String {
     BigInt::from_signed_bytes_be(bytes).to_str_radix(16)
 }
 
+fn offline_uuid(username: &str) -> anyhow::Result<Uuid> {
+    Uuid::from_slice(&Sha256::digest(username)[..16]).map_err(Into::into)
+}
+
 /// Login procedure for offline mode.
 fn login_offline(remote_addr: SocketAddr, username: String) -> anyhow::Result<NewClientInfo> {
     Ok(NewClientInfo {
         // Derive the client's UUID from a hash of their username.
-        uuid: Uuid::from_slice(&Sha256::digest(username.as_str())[..16])?,
+        uuid: offline_uuid(username.as_str())?,
         username,
         properties: vec![].into(),
         ip: remote_addr.ip(),
     })
 }
 
 /// Login procedure for BungeeCord.
-fn login_bungeecord(server_address: &str, username: String) -> anyhow::Result<NewClientInfo> {
+fn login_bungeecord(
+    remote_addr: SocketAddr,
+    server_address: &str,
+    username: String,
+) -> anyhow::Result<NewClientInfo> {
     // Get data from server_address field of the handshake
-    let [_, client_ip, uuid, properties]: [&str; 4] = server_address
-        .split('\0')
-        .take(4)
-        .collect::<Vec<_>>()
-        .try_into()
-        .map_err(|_| anyhow!("malformed BungeeCord server address data"))?;
+    let data = server_address.split('\0').take(4).collect::<Vec<_>>();
+
+    let ip = match data.get(1) {
+        Some(ip) => ip.parse()?,
+        None => remote_addr.ip(),
```

Comment:
it might be a good idea to put some comments to explain bungee's `ip-forward` and `online-mode` options